ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala

```diff
@@ -405,4 +405,35 @@ object domain {
     final case class CanActAs(party: Ref.Party) extends UserRight
     final case class CanReadAs(party: Ref.Party) extends UserRight
   }
+
+  sealed abstract class Feature extends Product with Serializable
+  object Feature {
+    case object UserManagement extends Feature
+    case object SelfServiceErrorCodes extends Feature
+    case object StaticTime extends Feature
+    case object ExperimentalOptionalLedgerId extends Feature
+
+    sealed case class CommandDeduplication(
+        offsetSupport: Option[CommandDeduplication.OffsetSupport.Value],
+        durationSupport: Option[CommandDeduplication.DurationSupport.Value],
+        deduplicationType: Option[CommandDeduplication.Type.Value],
+        maxDeduplicationDurationEnforced: Boolean,
+    ) extends Feature
+    object CommandDeduplication {
+      @SuppressWarnings(Array("org.wartremover.warts.Enumeration"))
+      object OffsetSupport extends Enumeration {
+        val OFFSET_NOT_SUPPORTED, OFFSET_NATIVE_SUPPORT, OFFSET_CONVERT_TO_DURATION = Value
+      }
+
+      @SuppressWarnings(Array("org.wartremover.warts.Enumeration"))
+      object DurationSupport extends Enumeration {
+        val DURATION_NATIVE_SUPPORT, DURATION_CONVERT_TO_OFFSET = Value
+      }
+
+      @SuppressWarnings(Array("org.wartremover.warts.Enumeration"))
+      object Type extends Enumeration {
+        val ASYNC_ONLY, ASYNC_AND_CONCURRENT_SYNC, SYNC_ONLY = Value
+      }
+    }
+  }
```

Comment:
I recommend factoring out everything up to this point into separate PR(s) sooner rather than later. Easier to review and you avoid having to fix conflicts over time.

Reply:
I'll split it out, as promised. I'm still getting used to the new workflow. We used to review by individual commit.